libraries/AP_RangeFinder/AP_RangeFinder_TOFSenseP_CAN.cpp

```diff
@@ -0,0 +1,58 @@
+#include "AP_RangeFinder_TOFSenseP_CAN.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include <AP_HAL/utility/sparse-endian.h>
+
+#if AP_RANGEFINDER_TOFSENSEP_CAN_ENABLED
+
+#include <AP_HAL/AP_HAL.h>
+
+RangeFinder_MultiCAN *AP_RangeFinder_TOFSenseP_CAN::multican_TOFSenseP;
+
+/*
+  constructor
+ */
+AP_RangeFinder_TOFSenseP_CAN::AP_RangeFinder_TOFSenseP_CAN(RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params) :
+    AP_RangeFinder_Backend_CAN(_state, _params)
+{
+    if (multican_TOFSenseP == nullptr) {
+        multican_TOFSenseP = new RangeFinder_MultiCAN(AP_CAN::Protocol::TOFSenseP, "TOFSenseP MultiCAN");
+        if (multican_TOFSenseP == nullptr) {
+            AP_BoardConfig::allocation_error("Rangefinder_MultiCAN");
+        }
+    }
+
+    {
+        // add to linked list of drivers
+        WITH_SEMAPHORE(multican_TOFSenseP->sem);
+        auto *prev = multican_TOFSenseP->drivers;
+        next = prev;
+        multican_TOFSenseP->drivers = this;
+    }
+}
+
+
+// handler for incoming frames. These come in at 10-30Hz
+bool AP_RangeFinder_TOFSenseP_CAN::handle_frame(AP_HAL::CANFrame &frame)
+{
+    WITH_SEMAPHORE(_sem);
+    const uint32_t id = frame.id - 0x200;
+
+    if (!is_correct_id(id)) {
+        return false;
+    }
+
+    const int32_t dist_cm = (int32_t)(frame.data[0] << 8 | frame.data[1] << 16 | frame.data[2] << 24) / 2560;
```

Comment:
Needs a comment about the unsigned conversion here. I  suspect this will work:
```
  const int32_t dist_cm = (int32_t)(frame.data[0] << 8U | frame.data[1] << 16U | frame.data[2] << 24U) / 2560;
```